docs/source/developer_guide/library_design.md

```diff
@@ -0,0 +1,111 @@
+# Library Design
+
+cuSpatial has two main components,
+the cuSpatial-python package and libcuspatial c++ package.
+This article dicusses the design of cuspatial-python,
+for information of libcuspatial,
+see [libcuspatial developer guide](link) and [c++ API reference](link).
+We refer to the python package as `cuspatial` and the c++ package as `libcuspatial` below.
+
+## Overview
+
+At a high level, `cuspatial` has three parts:
+- A GPU backed `GeoDataFrame` data structure
+- A set of computation APIs
+- A cython API layer
+
+## GPU Accelerated `GeoDataFrame` and `GeoSeries`
+
+Note: since the core data structure of cuspatial shares the same name as that
+of `geopandas`.
+We refer to geopandas' dataframe object as `geopandas.GeoDataFrame` and to
+cuspatial's dataframe object as `GeoDataFrame`.
+
+Under the hood,
+cuspatial can perform parallel computation on geometry data thanks to its [structure of array](https://en.wikipedia.org/wiki/Parallel_array) (SoA) format.
+Specifically,
+cuspatial adopts geoarrow format as the SoA standard.
+Geoarrow is a derived data type from the arrow list type adopting a [`Variable-size List Layout`](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout),
+with the inner-most layer storing the points with a `Fixed-size list layout` array with `size=2`.
+Per definition,
+each increase in geometry complexity (dimension, or multi-geometry) requires an extra level of indirection.
+Geoarrow allows a mixture of geometry types to present in the same column by adopting the [Dense Union Array Layout](https://arrow.apache.org/docs/format/Columnar.html#dense-union).
+In cusptial,
+we refer to each level of indirection from highest level to lowest by
+`geometries`, `parts`, `rings` and `coordinates`.
+The first three are integral offset arrays and the last is an floating-point interleaved xy-coordinated array.
+Read [geoarrow format documentation](https://github.com/geopandas/geo-arrow-spec/blob/main/format.md) specification for more detail.
+
+Cuspatial implements partial arrow union array via `GeoColumn` and `GeoMeta`.
+A GeoColumn is a composition of child columns and a `GeoMeta` object.
+The `GeoMeta` owns two arrays that are similar to the types buffer and offsets buffer from dense union.
```

Comment:
This paragraph is confusing. I can't figure out what it is trying to tell me.

What does this mean?: "Cuspatial implements partial arrow union array". I can't parse whether it partially implements an Arrow union array, or it fully implements a partial Arrow union array.  What is a "partial arrow union array", and why does cuSpatial need it?

```suggestion
cuSpatial implements partial Arrow union arrays via `GeoColumn` and
`GeoMeta`. A `GeoColumn` is a composition of child columns and a
`GeoMeta` object. The `GeoMeta` owns two arrays that are similar to the
types buffer and offsets buffer from Arrow dense union.
```

Reply:
> What does this mean?: "Cuspatial implements partial arrow union array". I can't parse whether it partially implements an Arrow union array, or it fully implements a partial Arrow union array. What is a "partial arrow union array", and why does cuSpatial need it?

I meant to say partially implements an arrow union array. Arrow union array type is more like a template - it allows arbitrary type composition. Cuspatial's geocolumn is a specialization of the union array that has 4 types: `point`, `multipoint`, `multilinestrings`, `multipolygons`.

Reply:
```suggestion
cuSpatial implements a specialization of Geoarrow dense union format via `GeoColumn` and
`GeoMeta`. A `GeoColumn` is a composition of child columns and a
`GeoMeta` object. The `GeoMeta` owns two arrays that are similar to the
types buffer and offsets buffer from Arrow dense union.
```